File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/servers/version/GivenAConnectionProviderThatDoesNotReturnProgramVersion/WhenReceivingThePromisedProgramVersion.java
package com.lasthopesoftware.bluewater.client.servers.version.GivenAConnectionProviderThatDoesNotReturnProgramVersion;

import com.lasthopesoftware.bluewater.client.connection.FakeConnectionProvider;
import com.lasthopesoftware.bluewater.client.connection.FakeConnectionResponseTuple;
import com.lasthopesoftware.bluewater.client.servers.version.ProgramVersionProvider;
import com.lasthopesoftware.bluewater.client.servers.version.SemanticVersion;
import com.lasthopesoftware.bluewater.shared.promises.extensions.FuturePromise;

import org.junit.BeforeClass;
import org.junit.Test;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import static org.assertj.core.api.Assertions.assertThat;

public class WhenReceivingThePromisedProgramVersion {

	private static SemanticVersion version;

	@BeforeClass
	public static void before() throws InterruptedException, TimeoutException, ExecutionException {
		final FakeConnectionProvider connectionProvider = new FakeConnectionProvider();

		connectionProvider.mapResponse((p) -> new FakeConnectionResponseTuple(200, "<Response Status=\"OK\"></Response>".getBytes()), "Alive");

		final ProgramVersionProvider programVersionProvider = new ProgramVersionProvider(connectionProvider);
		version = new FuturePromise<>(programVersionProvider.promiseServerVersion()).get(100, TimeUnit.MILLISECONDS);
	}

	@Test
	public void thenTheServerVersionIsNull() {
		assertThat(version).isNull();
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/servers/version/GivenAConnectionProviderThatDoesNotReturnProgramVersion/WhenReceivingThePromisedProgramVersion.kt
package com.lasthopesoftware.bluewater.client.servers.version.GivenAConnectionProviderThatDoesNotReturnProgramVersion

import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.connection.FakeConnectionProvider
import com.lasthopesoftware.bluewater.client.connection.FakeConnectionResponseTuple
import com.lasthopesoftware.bluewater.client.servers.version.LibraryServerVersionProvider
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.lasthopesoftware.promises.extensions.ProgressingPromise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class WhenReceivingThePromisedProgramVersion {
    private val version by lazy {
		val connectionProvider = FakeConnectionProvider()
		connectionProvider.mapResponse({
			FakeConnectionResponseTuple(
				200,
				"<Response Status=\"OK\"></Response>".toByteArray()
			)
		}, "Alive")
		val programVersionProvider = LibraryServerVersionProvider(mockk {
			every { promiseLibraryConnection(LibraryId(934)) } returns ProgressingPromise(connectionProvider)
		})
		programVersionProvider.promiseServerVersion(LibraryId(934)).toExpiringFuture().get()
	}

    @Test
    fun `then the server version is null`() {
        assertThat(version).isNull()
    }
}
